File: Day004_2025/Q004_GFG.py
class Solution:
    def countTriplets(self, arr, target):
        # Q. Count all triplets with given sum in sorted array
        ans = 0
        n = len(arr)
        for i in range(n - 2):
            j, k = i + 1, n - 1
            while j < k:
                triplet_sum = arr[i] + arr[j] + arr[k]
                if triplet_sum == target:
                    countJ, countK = 0, 0
                    valJ, valK = arr[j], arr[k]
                    while j <= k and arr[j] == valJ:
                        countJ += 1
                        j += 1
                    while j <= k and arr[k] == valK:
                        countK += 1
                        k -= 1
                    if valJ == valK:
                        ans += countJ * (countJ - 1) // 2
                    else:
                        ans += countJ * countK
                elif triplet_sum < target:
                    j += 1
                else:
                    k -= 1
        return ans


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        arr = list(map(int, input().split()))
        target = int(input())
        sol = Solution()
        print(sol.countTriplets(arr, target))
        print("~")

File: Day004_2025/Q004_GFG.java
import java.util.*;
import java.io.*;

class Solution {
    public int countTriplets(int[] arr, int target) {
        // Q. Count all triplets with given sum in sorted array
        int ans = 0;
        for (int i = 0; i < arr.length - 2; i++) {
            int j = i + 1;
            int k = arr.length - 1;
            while (j < k) {
                int sum = arr[i] + arr[j] + arr[k];
                if (sum == target) {
                    int countJ = 0, countK = 0;
                    int valJ = arr[j], valK = arr[k];
                    while (j <= k && arr[j] == valJ) {
                        countJ++;
                        j++;
                    }
                    while (j <= k && arr[k] == valK) {
                        countK++;
                        k--;
                    }
                    if (valJ == valK) {
                        ans += countJ * (countJ - 1) / 2;
                    } else {
                        ans += countJ * countK;
                    }
                } else if (sum < target) {
                    j++;
                } else {
                    k--;
                }
            }
        }
        return ans;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        while (t-- > 0) {
            String[] line = br.readLine().split(" ");
            int[] arr = Arrays.stream(line).mapToInt(Integer::parseInt).toArray();
            int target = Integer.parseInt(br.readLine());
            Solution sol = new Solution();
            System.out.println(sol.countTriplets(arr, target));
            System.out.println("~");
        }
    }
}
